src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderParser.cs

```diff
@@ -30,26 +31,40 @@ private sealed class ValueCreatorCache
             private static readonly FuncInstanceMethodInfo1<ValueCreatorCache, PrimitiveDataViewType, Func<RowSet, ColumnPipe>> _getCreatorVecCoreMethodInfo
                 = FuncInstanceMethodInfo1<ValueCreatorCache, PrimitiveDataViewType, Func<RowSet, ColumnPipe>>.Create(target => target.GetCreatorVecCore<int>);
 
-            private static volatile ValueCreatorCache _instance;
-            public static ValueCreatorCache Instance
+            private static volatile ValueCreatorCache _defaultInstance;
+            public static ValueCreatorCache DefaultInstance
             {
                 get
                 {
-                    return _instance ??
-                        Interlocked.CompareExchange(ref _instance, new ValueCreatorCache(), null) ??
-                        _instance;
+                    return _defaultInstance ??
+                        Interlocked.CompareExchange(ref _defaultInstance, new ValueCreatorCache(), null) ??
+                        _defaultInstance;
                 }
             }
 
+            private static readonly ConcurrentDictionary<DoubleParser.Options, ValueCreatorCache> _customInstances
+                = new ConcurrentDictionary<DoubleParser.Options, ValueCreatorCache>();
+
+            public static ValueCreatorCache GetInstanceWithDoubleParserOptions(DoubleParser.Options doubleParserOptions)
+            {
+                if (!_customInstances.ContainsKey(doubleParserOptions))
+                    return _customInstances.GetOrAdd(doubleParserOptions, new ValueCreatorCache(doubleParserOptions));
+
+                return _customInstances[doubleParserOptions];
```

Comment:
So I'm not super familiar with ConcurrentDictionary, but after looking at code in ML.NET that uses it, as well as some documentation, I think this is the way to go to avoid creating ValueCreatorCache instances more than needed, and only storing them once in the dictionary.

I did step through the code, while running the test I added, to make sure this worked correctly. And I also added some security checks on my test to confirm that only 1 instance is added to the dictionary.